sips/sip-021/sip-021-nakamoto.md

```diff
@@ -0,0 +1,1040 @@
+
+## Future Work: Transaction Replay on Bitcoin Forks
+
+The Bitcoin chain can fork.  This can be a problem, because Stacks transactions
+can be causally dependent on the now-orphaned Bitcoin state.  For example, any
+Stacks transaction that uses `(get-burn-block-info?)` may have a different
+execution outcome if evaluated after the Bitcoin block state from which it was
+mined no longer exists.
+
+To recover from Bitcoin forks, and the loss of data that may result, this
+proposal calls for dropping any previously-mined but now-invalid Stacks
+transactions from the Stacks chain history, but re-mining the set of Stacks
+transactions which remain valid across the Bitcoin fork in the same order in
+which they were previously mined.  That is, **transactions that were _not_
+causally dependent on lost Bitcoin state would remain confirmed on Stacks, in
+the same (relative) order in which they were previously mined.**
+
+To do so, Stackers would first observe that a Bitcoin fork has occurred, and
+vote on which Bitcoin block(s) were orphaned (even if it means sending the
+orphaned data to each other, since not all Stacks nodes may have seen it).  Once
+Stackers agree on the sequence of orphaned Bitcoin blocks, they identify which
+Stacks blocks would be affected.  From there, they each replay the affected
+Stacks blocks' transactions in the same order, and in doing so, identify which
+transactions are now invalid and which ones remain valid.  Once they have this
+subsequence of still-valid transactions, they advertise it to miners, and only
+sign off on Stacks blocks that include a prefix of this subsequence that has not
+yet been re-mined (ignoring `Coinbase`, `TenureChange`, and `TenureExtension`
+transactions).  This way, the Stacks miners are compelled to replay the
+still-valid Stacks transactions in their same relative order, thereby meeting
+this guarantee.
+
+Importantly, this transaction replay feature is directed exclusively by Stacker
+and miner policy logic.  It is not consensus-critical, and in fact cannot be
+because not all miners or Stackers may have even seen the orphaned Bitcoin state
+(which precludes them from independently identifying replay transactions; they
+must instead _work together_ to do so off-chain).  Therefore, this feature's
+implementation can be deferred until after this SIP is ratified.
+
+# Backwards Compatibility
+
+This proposal is a breaking change.  However, all smart contracts published
+prior to this proposal's activation will be usable after this proposal
+activates.  Of particular note is that the `block-height` Clarity variable will
+increment at a much faster pace, since blocks will be produced much more
+quickly.
+
+A separate SIP will be written to describe any changes to Clarity that may take
+effect with the activation of this SIP.
+
+Some design decisions of this SIP were undertaken to support sBTC, which is
+described in a separate SIP.  In particular, the sBTC system requires a notion
```

Comment:
[todo] link to sbtc SIP?

Reply:
Not in-scope.  sBTC needs Nakamoto, but Nakamoto does not need sBTC.

Reply:
I don't follow: you already allude to 'sBTC, which is described in a separate SIP' - wouldn't linking to that SIP provide more context to the reader?

Reply:
I'm reluctant to have this SIP depend on the sBTC SIP if at all possible, because there's no guarantee that sBTC activates along with Nakamoto.  sBTC may even undergo significant revisions between Nakamoto's activation and sBTC's activation.  I'd almost rather drop this sentence, and instead describe how Nakamoto makes it easier to represent Bitcoin state on Stacks (which sBTC needs, but is by itself beneficial for other tech such as Stacks-wrapped ordinals).